src/span.cpp

```diff
@@ -14,49 +14,50 @@ const std::string datadog_resource_name_tag = "resource.name";
 const std::string datadog_service_name_tag = "service.name";
 }  // namespace
 
-Span::Span(std::shared_ptr<const Tracer> tracer, std::shared_ptr<SpanBuffer<Span>> buffer,
+SpanData::SpanData(std::string type, std::string service, ot::string_view resource,
+                   std::string name, uint64_t trace_id, uint64_t span_id, uint64_t parent_id,
+                   int64_t start, int64_t duration, int32_t error)
+    : type(type),
+      service(service),
+      resource(resource),
+      name(name),
+      trace_id(trace_id),
+      span_id(span_id),
+      parent_id(parent_id),
+      start(start),
+      duration(duration),
+      error(error) {}
+
+SpanData::SpanData() {}
+
+uint64_t SpanData::traceId() const { return trace_id; }
+uint64_t SpanData::spanId() const { return span_id; }
+
+std::unique_ptr<SpanData> makeSpanData(std::string type, std::string service,
+                                       ot::string_view resource, std::string name,
+                                       uint64_t trace_id, int64_t span_id, uint64_t parent_id,
+                                       int64_t start) {
+  return std::unique_ptr<SpanData>{
+      new SpanData(type, service, resource, name, trace_id, span_id, parent_id, start, 0, 0)};
+}
+
+std::unique_ptr<SpanData> stubSpanData() { return std::unique_ptr<SpanData>{new SpanData()}; }
+
+Span::Span(std::shared_ptr<const Tracer> tracer, std::shared_ptr<SpanBuffer> buffer,
            TimeProvider get_time, uint64_t span_id, uint64_t trace_id, uint64_t parent_id,
            SpanContext context, TimePoint start_time, std::string span_service,
-           std::string span_type, std::string span_name, ot::string_view resource,
-           const ot::StartSpanOptions &options)
+           std::string span_type, std::string span_name, ot::string_view resource)
     : tracer_(std::move(tracer)),
       buffer_(std::move(buffer)),
       get_time_(get_time),
-      span_id(span_id),
-      trace_id(trace_id),
-      parent_id(parent_id),
       context_(std::move(context)),
       start_time_(start_time),
-      service(span_service),
-      type(span_type),
-      name(span_name),
-      resource(resource),
-      error(0),
-      start(std::chrono::duration_cast<std::chrono::nanoseconds>(
-                start_time_.absolute_time.time_since_epoch())
-                .count()),
-      duration(0) {
-  buffer_->registerSpan(*this);
-}
-
-Span::Span(Span &&other)
-    : tracer_(other.tracer_),
-      buffer_(other.buffer_),
```

Comment:
To fix the bug, literally all I needed to do was comment out these two preceding lines. If this move constructor is called because of destruction and the Span being moved away to the buffer for writing, then if the tracer/buffer have also destructed then these pointers segfault.

Reply:
But I didn't do the simple fix, I did something that prevents us from keeping ANY pointers (or, er, point _less_ data) around after Span finish.